### PR TITLE
Add events lag metric

### PR DESCRIPTION
**Pull request by lvoloshyn-sekoia**

### CODE REVIEW

File: PandaSecurity/aether_endpoint_security_api/trigger_security_events.py [hunk 1]
```diff
@@ -95,6 +108,24 @@ def _fetch_events(self) -> None:
...
 
+                OUTCOMING_EVENTS.labels(account_id=self.module.configuration["account_id"], type=event_type_name).inc(
+                    len(message_batch)
+                )
+
+            # get the ending time and compute the duration to fetch the events
+            batch_end_time = time.time()
+            batch_duration = int(batch_end_time - batch_start_time)
+            FORWARD_EVENTS_DURATION.labels(
+                account_id=self.module.configuration["account_id"], type=event_type_name
```
Comment (squioc) [inline PandaSecurity/aether_endpoint_security_api/trigger_security_events.py:119]:
```suggestion
                type=event_type_name
```

File: PandaSecurity/aether_endpoint_security_api/trigger_security_events.py [hunk 2]
```diff
@@ -95,6 +108,24 @@ def _fetch_events(self) -> None:
...
+            ).observe(batch_duration)
+
+            # compute the events lag
+            last_message_timestamp = datetime.strptime(last_message_date, self.RFC3339_STRICT_FORMAT)
+            events_lag = (datetime.utcnow() - last_message_timestamp).total_seconds()
+            EVENTS_LAG.labels(account_id=self.module.configuration["account_id"], type=event_type_name).observe(
```
Comment (squioc) [inline PandaSecurity/aether_endpoint_security_api/trigger_security_events.py:125]:
```suggestion
            EVENTS_LAG.labels(type=event_type_name).observe(
```

File: VadeSecure/vadesecure_modules/metrics.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    namespace=prom_namespace,
+    labelnames=["tenant_id", "type"],
+)
+
+EVENTS_LAG = Histogram(
+    name="events_lags",
+    documentation="The delay, in seconds, from the date of the last event",
+    namespace=prom_namespace,
+    labelnames=["tenant_id", "type"],
+)
```
Comment (squioc) [inline VadeSecure/vadesecure_modules/metrics.py:35]:
```suggestion
INCOMING_MESSAGES = Counter(
    name="collected_messages",
    documentation="Number of messages consumed",
    namespace=prom_namespace,
    labelnames=["type"],
)

# Declare common prometheus metrics
prom_namespace = "symphony_module_common"

OUTCOMING_EVENTS = Counter(
    name="forwarded_events",
    documentation="Number of events forwarded to Sekoia.io",
    namespace=prom_namespace,
    labelnames=["type"],
)

FORWARD_EVENTS_DURATION = Histogram(
    name="forward_events_duration",
    documentation="Duration to collect and forward events",
    namespace=prom_namespace,
    labelnames=["type"],
)

EVENTS_LAG = Histogram(
    name="events_lags",
    documentation="The delay, in seconds, from the date of the last event",
    namespace=prom_namespace,
    labelnames=["type"],
)
```

File: VadeSecure/vadesecure_modules/trigger_m365_events.py [hunk 4]
```diff
@@ -143,11 +147,32 @@ def _fetch_events(self) -> None:
                     level="info",
                 )
 
+                INCOMING_MESSAGES.labels(tenant_id=self.configuration.tenant_id, type=event_type).inc(
```
Comment (squioc) [inline VadeSecure/vadesecure_modules/trigger_m365_events.py:150]:
```suggestion
                INCOMING_MESSAGES.labels(type=event_type).inc(
```

File: VadeSecure/vadesecure_modules/trigger_m365_events.py [hunk 5]
```diff
@@ -143,11 +147,32 @@ def _fetch_events(self) -> None:
...
+                    len(message_batch)
+                )
+
+                last_message = message_batch[-1]
+                last_message_date = self._get_last_message_date([last_message])
+                events_lag = int(time.time() - last_message_date.timestamp())
+                EVENTS_LAG.labels(tenant_id=self.configuration.tenant_id, type=event_type).observe(events_lag)
```
Comment (squioc) [inline VadeSecure/vadesecure_modules/trigger_m365_events.py:157]:
```suggestion
                EVENTS_LAG.labels(type=event_type).observe(events_lag)
```

File: VadeSecure/vadesecure_modules/trigger_m365_events.py [hunk 6]
```diff
@@ -143,11 +147,32 @@ def _fetch_events(self) -> None:
...
+
                 for emails in self._chunk_events(list(message_batch), self.max_batch_size):
                     self._send_emails(
                         emails=list(emails),
                         event_name=f"M365-events_{last_message_date}",
                     )
+
+                OUTCOMING_EVENTS.labels(tenant_id=self.configuration.tenant_id, type=event_type).inc(
```
Comment (squioc) [inline VadeSecure/vadesecure_modules/trigger_m365_events.py:165]:
```suggestion
                OUTCOMING_EVENTS.labels(type=event_type).inc(
```

File: VadeSecure/vadesecure_modules/trigger_m365_events.py [hunk 7]
```diff
@@ -143,11 +147,32 @@ def _fetch_events(self) -> None:
...
+                    len(message_batch)
+                )
+
+            # get the ending time and compute the duration to fetch the events
+            batch_end_time = time.time()
+            batch_duration = int(batch_end_time - batch_start_time)
+            FORWARD_EVENTS_DURATION.labels(tenant_id=self.configuration.tenant_id, type=event_type).observe(
```
Comment (squioc) [inline VadeSecure/vadesecure_modules/trigger_m365_events.py:172]:
```suggestion
            FORWARD_EVENTS_DURATION.labels(type=event_type).observe(
```

**Review by squioc — APPROVED** [review]
Nice job.
I made some suggestions.

File: Sophos/sophos_module/trigger_sophos_edr_events.py [hunk 8]
```diff
@@ -124,6 +124,15 @@ def get_next_events(self, cursor: str | None) -> dict | None:
 
         return response.json()
 
+    def _get_most_recent_timestamp_from_items(self, items: list[dict]):
+        def _extract_timestamp(item: dict) -> float:
+            RFC3339_STRICT_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+            return datetime.datetime.strptime(item["created_at"], RFC3339_STRICT_FORMAT).timestamp()
+
+        latest_message = max(items, key=lambda item: _extract_timestamp(item))
+        latest_message_timestamp = _extract_timestamp(latest_message)
```
[the remark this answers is not in the document]
Comment (lvoloshyn-sekoia) [reply Sophos/sophos_module/trigger_sophos_edr_events.py:133]: Great idea!

File: Duo/duo/connector.py [hunk 9]
```diff
@@ -135,6 +135,21 @@ def fetch_batches(self):
         for events in self.get_events_iterator():
             batch_start_time = time.time()
 
+            if len(events) > 0:
+                if self._log_type == LogType.TELEPHONY:
+                    # Telephony logs have their datetime represented as "2023-03-21T22:34:49.466370+00:00"
+                    most_recent_event = max(events, key=lambda item: item["ts"])
+                    most_recent_timestamp = datetime.datetime.fromisoformat(most_recent_event["ts"]).timestamp()
+                else:
+                    most_recent_event = max(events, key=lambda item: item.get("timestamp"))
```
Comment (squioc) [inline Duo/duo/connector.py:144]: If one of the events doesn't contain the field `timestamp`, the max function will generate an exception. I suggest filtering the list of events before.
```suggestion
                    most_recent_event = max(filter(lambda item: item.get("timestamp"), events))
```

**Review by squioc — CHANGES_REQUESTED** [review]

**Review by squioc — APPROVED** [review]

**Pull request merged**